jfr/src/main/java/io/netty/handler/jfr/JfrEventHandler.java

```diff
@@ -0,0 +1,375 @@
+/*
+ * Copyright 2012 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.jfr;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandler.Sharable;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import jdk.jfr.Category;
+import jdk.jfr.Enabled;
+import jdk.jfr.Event;
+import jdk.jfr.Label;
+import jdk.jfr.Name;
+
+import java.net.SocketAddress;
+
+/**
+ * A {@link ChannelHandler} that creates JFR events for all Netty events/operations using the JDK Flight Recorder using
+ * the jdk.jfr APIs directly.
+ * <p>
+ * Added for benchmark comparisons.
+ */
+@Sharable
+@SuppressWarnings({ "StringConcatenationInsideStringBufferAppend", "StringBufferReplaceableByString" })
+public class JfrEventHandler extends ChannelDuplexHandler {
+    // TODO: Check enablement, can we ever be sure not to enable this?
+
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+        ChannelRegisteredEvent event = new ChannelRegisteredEvent();
+        event.begin();
+        ctx.fireChannelRegistered();
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    @Override
+    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
+        ChannelUnregisteredEvent event = new ChannelUnregisteredEvent();
+        event.begin();
+        ctx.fireChannelUnregistered();
```

Comment:
generally speaking the `ctx.*` methods should never throw but would still make sense to use a `try {...} finally {...}` block ? Same comment is true for all the methods

Reply:
I guess then you would also want to mark the event as completing exceptionally?